free-1.md

```diff
@@ -57,12 +115,48 @@ making language design decisions), as opposed to "systems" aspects of the
 project (e.g., implementing a complicated semantics that doesn't require a lot
 of language design)?_
 
+Probably the majority of the time would be language-related. The implementation
+is relatively simple (assuming it isn’t cross-platform); the API to create pipes
+already exists in the `pipe(2)` system call. Other potential features get more
+complicated, but the system calls are still all relatively simple and the main
+part of the work would be the language.
+
+One potential pitfall would be that it might take a while to reimplement
+existing shell syntax. A minimal Bourne-compatible shell would probably be
+almost trivial to implement, but adding the features that users of `bash` and
+`zsh` are accustomed to would take forever before further language design even
+became an issue. If this can be implemented as an internal DSL within shell
+scripting, or as an extension to `zsh` (which I know to be extensible, although
+I know nothing about the extension mechanisms), then this pitfall is avoided.
 
 ### Scope
 _How big an idea is this? How ambitious is this project?_
 
+That’s hard to tell. A very simple, usable prototype could probably be completed
+in a few hours, which is obviously too small. There are a lot of features it
+would be nice to include (expanding to sockets instead of just pipes, adding
+I/O via tmpfs files instead of pipes, adding wrappers around commands to make
+them work more uniformly or programmatically, etc.), but it’s difficult to know
+whether any of those features are integrable into the language in a reasonable
+way.
+
+I expect that, if a language feature could be developed for wrappers (similar
+to the existing means of defining functions in Bourne-like shells), then the
+project would have a potentially large scope. I’m not sure about how feasible it
+would be to integrate such a method with existing `sh`-like syntax, though.
 
 ### Benefits and drawbacks
 _Why might this be a good idea for a project? Why might this not be a good idea 
 project?_
 
+This could be a good idea for a project because, assuming the issues in the
+previous two sections were resolved, there is a lot of freedom for language
+design. There would relatively little difficulty in implementing the language,
+and the language could be built in an agile way so that there was a working
+deliverable product at the end of the semester regardless of potential poor
+estimation of scope.
+
+If the problems of the last two sections (not reimplementing a full-featured
+shell; allowing command wrappers for uniformity) were not solved or otherwise
+avoided, then the project would be either far too large in scope or far too
+small in scope.
```

Comment:
This idea seems interesting. If you have a few, specific examples of programs-as-DAGs, then I would say you could proceed with this project. But if you don't have more than one specific example, then you risk building something that may not be useful.

Also, if you were to go with this project, I recommend starting with commands, variables, and graphs. Once you've worked out those three features, then move on to other control-flow structures.

Would it be possible to implement these ideas as an internal DSL in an existing shell?